### ISSUE TITLE
Is there a way to avoid resending all events after restarting the container?

### ISSUE THREAD

**Issue by hoegertn** [kelseyhightower/journal-2-logentries#6]
On container restart all event in journal are resend to logentries. Maybe this could be prevented by using the range header.

**Comment by kelseyhightower** [comment]
@hoegertn Yes, we need to use the features of the systemd journal. We can start pulling logs based on time, or something else.

**Comment by tpett** [comment]
This isn't the ideal solution, but I've created a fork that limits entries to the last boot. See this commit: https://github.com/watermarkchurch/journal-2-logentries/commit/9d6aa3a7a833f433f52e624c4aaea784c16f5533

It is published as a container at watermarkchurch/journal-2-logentries

The Range header that @hoegertn suggested seems like it would be a better fit, but I couldn't figure out how to make it work without an explicit reference to a log cursor (which the container won't have). That being said I suppose if we pass a cursor in through an environment variable on container boot with something like `journalctl -show-cursor -r -n 1` we might be able to make it work. For now though, the boot flag will work for me.

**Comment by jaroslavl1** [comment]
Actually sending logs to logentries is quite easy if you create a systemd unit that uses journalctl, awk and ncat to send logs.

Incomplete example for CoreOS:

```
ExecStart=/bin/bash -c "journalctl --since \"$(etcdctl get /logentries.com/$(hostname)/lastlog)\" -o json --boot -f | awk '{ print \"${LOG_KEY}\",$0; fflush(); }' | ncat --ssl api.logentries.com 20000"
```

You will also need to handle setting the etcd record in case it isn't set yet and set it when service stops.

If you want less junk in logentries then also use jq before awk. LOG_KEY is the log key assigned to you by logentries.

**Comment by sweber83** [comment]
@jaroslavl1 
Thanks for pointing out this simple method.
I build this into a unit file for CoreOS:

```
[Unit]
Description=Forward Systemd Journal to logentries.com
After=systemd-journald.service

[Service]
Restart=always
IgnoreSIGPIPE=no
ExecStart=/usr/bin/bash -c "\
  token=$(etcdctl get /logentries.com/token);\
  date=$(etcdctl get /logentries.com/$(hostname)/lastlog);\
  if [ -z \"$date\" ]; then\
    date=$(date +%F);\
  fi;\
  journalctl --since \"$date\" -o json --boot -f |\
  jq -c '{timestamp:.__REALTIME_TIMESTAMP,time:.__REALTIME_TIMESTAMP | tonumber | ( . / 1000000 ) | strftime(\"%Y-%%m-%dT%%H:%M:%SZ\"),host:._HOSTNAME,unit:._SYSTEMD_UNIT,message_raw: .MESSAGE, message_parsed: ( try ( .MESSAGE | fromjson ) catch false)} | {timestamp:.timestamp, time:.time, host:.host, unit:.unit, message:(if .message_parsed == false then .message_raw else .message_parsed end)}' |\
  awk -v token=\"$token\" '{ print token,$0; fflush(); }' |\
  ncat --ssl api.logentries.com 20000;"

ExecStop=/usr/bin/bash -c "etcdctl set /logentries.com/$(hostname)/lastlog \"$(date +'%F %%H:%M:%S')\""

[X-Fleet]
Global=true
```

It is really important to set `IgnoreSIGPIPE=no`, otherwise it can happen that ncat exits and the rest of the pipeline continues running.
The jq program formats the timestamp with strftime and tries to parse the message as json, so that one can use json logging from nginx for example.
